ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -139,33 +131,29 @@ public struct SignalProducer<Value, Error: ErrorType> {
 			// Assigned to when replay() is invoked synchronously below.
 			var token: RemovalToken?
 
-			let replay: () -> () = {
-				let originalState = state.modify { state in
-					var mutableState = state
-					token = mutableState.observers?.insert(observer)
-					return mutableState
-				}
-
-				for value in originalState.values {
-					observer.sendNext(value)
-				}
+			// This lock is used to ensure that each observer receives events
+			// in the same order as they were buffered.
+			// This is required because another thread may push an event while
+			// the buffer is being replayed to the observer.
+			// In that case the lock prevents the observer from receiving that event
+			// before it receives all of the buffered events.
+			let replayLock = NSLock()
+			replayLock.name = "org.reactivecocoa.ReactiveCocoa.SignalProducer.buffer.replay"
+			replayLock.lock()
```

Comment:
State is per entire buffer. The idea of this fix is to prevent the deadlock by having an individual lock for each observer, and using that lock to serialize sending the events to that observer only.